### DISCUSSION TITLE
Universe Creator

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#34] [Ideas]
This is a proposal for the universe creator, which is the tool used to create:

- Solar Systems (Thorium Nova Prime called them "planetary systems" because that's more specific, but it's also a mouthful, so we're calling it the slightly less accurate but easier to understand "solar systems")
- Stars
- Planets
- Moons
- Permanent "ships", like space stations and satellites.

Universes are persisted in plugins (https://github.com/Thorium-Sim/thorium-nova/discussions/26).

Note that the scope of this proposal is _just_ the universe creator tool which is used to edit the universe in the plugin, not tools used during a flight by the crew or Flight Director.

For Thorium Nova Prime, kept track of the features using my own project management system. Here's all of the individual tasks I listed there for universe creation.

- Distances and sizes at interstellar and solar system level are accurate to real-life.
- Two separate levels of detail - interstellar-level for all solar systems, and system-level for all objects within an individual solar system.
- Create solar systems
  -  How to position these in 3D space - place at a certain distance from the camera.
  - Edit the position of solar systems in 3D space - drag along a plane a certain distance from the camera. 
- Create stars
  - Support binary/trinary stars
  - Stars always exist at the center of the solar system
- Select objects
- Delete objects
- Edit object properties
  -  Properties for solar systems, stars, planets, and ships are different. Pay attention to that.
- Find the distance between objects - LY in the interstellar level, KM in solar system level.
- Add planets to solar system based on orbit radius, angle, and azimuth (which means planets can't just be placed anywhere in the solar system - only along that orbital path based on the location of the star)
  - Planet textures for surface, clouds, and rings

**Excellent experience editing the universe**
- Pan and zoom (implemented really well)
- Have an option to flatten the Y axis (since this messes with pan/zoom)
- Draggable Zoom Slider + zoom with pinch to zoom or holding down Option/Alt key. 
- Button to center the camera on the selected object.

Feedback, additions, and suggestions welcome!

**Comment by astrolamb-gaming** [comment]
Solid plan. I really like the idea of having realistic planetary systems. It'll add a bit of complexity for plotting courses and such, and will also prevent systems from seeming too static since they'll be somewhat differently arranged each time a crew visits the system. 

A couple things to consider which aren't exactly related to the universe editor itself: 

What sublight speeds will a ship be able to achieve? How long would it take to travel between earth and mars at the closest point in their orbits, for example? 

How will warp and other ftl systems work? In artemis a pilot can completely change course while at warp. Would in-system warping be possible? For thorium i imagine selecting a destination and aligning to the appropriate course before engaging warp, and not being able to deviate without dropping out of warp. Is this consistent with your vision?

**Reply by astrolamb-gaming** [reply]
Will gravity from celestial objects effect the ship's course? (I kinda want it to, but it may increase the complexity of several things)

**Reply by alexanderson1993** [reply]
> What sublight speeds will a ship be able to achieve? How long would it take to travel between earth and mars at the closest point in their orbits, for example?

I address this in this blog post: https://thoriumsim.com/blog/distances-engines-and-speeds

tl;dr - you would use Warp speed to travel between planets with an autopilot automatically cutting power when you reach your destination (since it's so fast). At the 1,500 km/s max sublight speed I mention there, it would take 10 hours to react Mars from Earth, far too long for the stories we want to tell.

Specific mechanics I want to build into the Warp drive are also mentioned in that article. Basically, it takes some time to charge up the warp drive before you activate it, which means you can't just turn it on and off on a whim - you have to be deliberate about it. I don't see any reason why the course can't change at warp speed though - it's just rotating the ship using the thrusters after all. You just turn like a boat.

Of course, a wormhole or jump drive-type engine would have totally different mechanics.

> they'll be somewhat differently arranged each time a crew visits the system

Planets orbit so slowly that it's unlikely this will be the case. In fact, I had little intention originally of simulating the movement of planets around their star(s), but...

> Will gravity from celestial objects effect the ship's course?

It's actually not that big of a difference. Every object will have a mass property, so calculating the acceleration due to gravity isn't much effort. It can be done as part of a separate ECS system. The effect of gravity will be totally minuscule - likely imperceptible - but there's no reason not to add it if we want. We could even use it to simulate the movement of planets and such, although there are performance gains to be had by assuming planets don't move at all.

**Reply by astrolamb-gaming** [reply]
> The effect of gravity will be totally minuscule - likely imperceptible - but there's no reason not to add it if we want. We could even use it to simulate the movement of planets and such, although there are performance gains to be had by assuming planets don't move at all.

Orbits are a pretty big deal when you're close to a planet. This is minimized by the speeds a ship is capable of given your stated specs, but if in close orbit, it's noticeable. But I'm ok with not bothering with this. 

As for movement of planets, i would not recommend live simulation of movement. However, if there's an in-game month that goes by between visits to a particular system, i think that system's bodies should reflect that difference. It's a simple enough calculation (especially assuming 2-body calculations, i.e. sun and planet or planet and moon) that could be performed wherever entering a system.

**Reply by Froginator45** [reply]
I really enjoy the idea of realistic planetary systems. But, would anyone like to have the option to create planetary systems that don't follow physics for a specific mission? i.e. some alien race created something that is creating a space that works in an alternate physical domain or built something so massive that it has become the center of gravity instead of the local sun? Just thinking of cases where someone might want an override button.

**Reply by alexanderson1993** [reply]
This seems a little niche for what we're trying to accomplish, and would likely cover < 1% of use cases. 

Part of the reason why I like defining planet's positions by their orbit is that it makes it _really_ easy to draw orbit lines, which is really nice for quickly identifying which planet is which, how many planets there are, etc. especially when planets are so small at an interplanetary level.

**Comment by aBlueShadow** [comment]
> Of course, a wormhole or jump drive-type engine would have totally different mechanics.

Actually, it doesn't have to be that different. Also technically, both drives are usually explained by warping spacetime.

You mentioned that interplanetary warp using interstellar speed would let you reach other planets within the blink of an eye. So, for an observer, this would basically look like a jump drive. Of course, stopping could not be done manually, you have to instruct the computer beforehand when it should shut down the engine exactly, and instead of setting the amount of split-seconds, you would conveniently enter the distance, while the computer will do the math.
EmptyEpsilon might serve as an example: you set the direction like you would do for warp and impulse, by steering the ship, but additionally set a jump distance.

So lorewise, you could always use the same drive and speeds, but within planetary systems, the drive changes to a "jump mode"

**Reply by alexanderson1993** [reply]
First off, hi @aBlueShadow. 👋 Thanks for popping by!

For me, it's a question of whether you traverse space linearly from one point to another, or whether you pop out of one area of space and reappear in a different area of space. The mechanics for that are slightly different.

Another difference is that of storytelling. I'm a firm believer that travel time is an important element of the story, an instant travel should generally be avoided. For Jump Drives, that means that between disappearing and reappearing, there is a bit of time where the ship is just... nowhere, until the Flight Director or some automated trigger makes them pop back into reality.

**Reply by aBlueShadow** [reply]
But can't a longer wind up time for the drive serve a similar purpose, story-telling-wise? In world explanation could be, that preparation would be longer, because the calculations have to be very precise inside of the solar system. BTW that is actually meant as a question, I am not quiet sure if it really comparable. Technically it would be almost the same, as it is a transition phase that could be used for down-time, character play, or be interrupted for action. But it might still feel different.

**Reply by alexanderson1993** [reply]
The way I see it, with a longer wind up time before the jump, the ship is still in danger. They could be ambushed or something else wrong could happen.

Once they jump, they are home free and can breathe a sigh of relief and relax.

**Reply by aBlueShadow** [reply]
Ah I understand, so you see it as a safe place, not a journey that can be interrupted by encounters "on the road".

**Reply by alexanderson1993** [reply]
Correct, and this is the case for both jump drive and warp engines. Realistically speaking, being interrupted by an encounter while traveling magnitudes past the speed of light is... unlikely. (Unless those encounters occur because of shenanigans happening inside the ship)

**Reply by aBlueShadow** [reply]
True. Still a common trope though, and there are a number of episodes where the ship is suddenly falling out of warp/hyperspace/whatever :-). But yeah, an ambush/track down during preparation time is probably still much more likely

**Reply by alexanderson1993** [reply]
And that preparation time is definitely something I want to have. In [this blog post](https://thoriumsim.com/blog/distances-engines-and-speeds) I talk about how the Warp Dynamo needs to spool before you can jump to warp, so you're not just warping whenever you feel like. Same thing could apply to the Jump Drive.

Having a build up of tension and anxiety as you prep the engines is as important as the relief that comes when you are finally underway.
